pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package remoteconfighandler
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state/products/apmsampling"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/log"
+	"github.com/DataDog/datadog-agent/pkg/trace/sampler"
+)
+
+type apmsamplingRemoteConfigID string
+
+const (
+	dynamicRatesByService     apmsamplingRemoteConfigID = "dynamic_rates"
+	userDefinedRatesByService apmsamplingRemoteConfigID = "user_rates"
+	samplerconfig             apmsamplingRemoteConfigID = "samplerconfig"
+)
+
+type prioritySampler interface {
+	UpdateTargetTPS(targetTPS float64)
+	UpdateRemoteRates(updates []sampler.RemoteRateUpdate)
+}
+
+type errorsSampler interface {
+	UpdateTargetTPS(targetTPS float64)
+}
+
+type rareSampler interface {
+	SetEnabled(enabled bool)
+}
+
+// RemoteConfigHandler holds pointers to samplers that need to be updated when APM remote config changes
+type RemoteConfigHandler struct {
+	remoteClient          config.RemoteClient
+	prioritySampler       prioritySampler
+	errorsSampler         errorsSampler
+	rareSampler           rareSampler
+	agentConfig           *config.AgentConfig
+	remoteConfigPathRegex *regexp.Regexp
+}
+
+func New(conf *config.AgentConfig, prioritySampler prioritySampler, rareSampler rareSampler, errorsSampler errorsSampler) *RemoteConfigHandler {
+	if conf.RemoteSamplingClient == nil {
+		return nil
+	}
+
+	// paths are in the form datadog/<org ID>/<product name>/<config ID>/config
+	r := regexp.MustCompile(`datadog\/\d+\/APM_SAMPLING\/(.+)\/config`)
+
+	return &RemoteConfigHandler{
+		remoteClient:          conf.RemoteSamplingClient,
+		prioritySampler:       prioritySampler,
+		rareSampler:           rareSampler,
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
```

Comment:
Should we log the error? Or it's logged by the remote client?

Reply:
I don't understand, I am logging it on line 81, right?

Reply:
Sorry, my bad.